ENH: Add wrapper for ?pteqr

mdhaber opened the pull request:
This PR adds Python level wrappers for the LAPACK functions ?pteqr as part of a NumFocus small development grant. ?pteqr computes the eigenvectors and eigenvalues of a positive definite tridiagonal matrix. SciPy has wrappers tailored for eigenvalue problems involving other classes of matrices, and positive definite tridiagonal matrices are important matrix class, so SciPy would benefit from enhanced support for tridiagonal matrices.

This closes #11224

Thanks to @Kai-Striega for writing the wrapper (I'm just creating the PR) and @swallan for writing the tests.

mdhaber commented:
@ilayn I saw your conversation with @Kai-Striega in gh-11224, and I believe I followed. I understand that when a char is acting as a switch you typically convert to integers, there are three different values here. How would the user know how the integers map to the characters?

I took a look at other examples of `compv` and similar and I tried to adapt that approach to determine the value for ldz. However, I think it would be better to have some sort of check to ensure for `z.shape[0] ==n` when `compz!='N'`. The way it is now, if `z.shape[0]<n`, the code returns an incorrect result rather than immediately raising an exception, causing `test_pteqr_raise_error_wrong_shape` to fail.

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1957,6 +1957,141 @@ def test_geqrfp_lwork(dtype, shape):
+def test_pteqr_NAG_f08jgf(compz, d, e, d_expect, z_expect):
+    '''
+    Implements real (f08jgf) example from NAG manual.
+    https://www.nag.com/numeric/fl/nagdoc_latest/html/f08/f08jgf.html
```
I can remove the link if necessary.

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1957,6 +1957,141 @@ def test_geqrfp_lwork(dtype, shape):
+                         zip(DTYPES, REAL_DTYPES + REAL_DTYPES))
+@pytest.mark.parametrize("compz", ["I", "N", "V"])
+def test_pteqr_error_singular(dtype, realtype, compz):
+    np.random.seed(42)
```
`seed` absolute import.

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1957,6 +1957,141 @@ def test_geqrfp_lwork(dtype, shape):
+    correct eigenvalues with scipy.linalg.eig. With applicable compz=I it
+    tests that z can reform A.
+    '''
+    np.random.seed(42)
```
`seed` absolute import here and below.

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1957,6 +1957,141 @@ def test_geqrfp_lwork(dtype, shape):
+    '''
+    np.random.seed(42)
+    rtol = 2500*np.finfo(dtype).eps
+    atol = 1000*np.finfo(dtype).eps
```
Only `atol` needed.

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1957,6 +1957,141 @@ def test_geqrfp_lwork(dtype, shape):
+        A_eig = A_eig + np.diag(np.zeros(n) + 4*n)
+        A_eig = (A_eig + A_eig.conj().T) / 2
+        # obtain right eigenvectors (orthogonal)
+        vr = eig(A_eig)[1]
```
```suggestion
        vr = eigh(A_eig)[1]
```
here and elsewhere

tylerjereddy commented:
Note the Azure pipelines have stopped triggering for NumPy (https://github.com/numpy/numpy/pull/15747) and SciPy (here).

tylerjereddy commented:
Ah, maybe just the merge conflicts actually

tylerjereddy commented:
/azp run

ilayn commented:
>  How would the user know how the characters map to integers?

They actually don't but that's kind of what the standard is. 0, 1, 2 is used throughout the eig wrappers for any switches and also the variable name template is `compute_x` where `x` is the variable name.

mdhaber commented:
Renamed `compute_z` and hid `work`.

@ilayn Do you agree that we also need some sort of check to ensure that `z.shape[0] ==n` when `compz!='N'`. The way it is now, if `z.shape[0]<n`, the code returns an incorrect result, whereas I think it should immediately raise an exception.

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -2081,6 +2081,140 @@ def test_pttrf_pttrs_NAG(d, e, d_expect, e_expect, b, x_expect):
+    assert_raises(ValueError, pteqr, d, e[:-1], z=z, compz=compz)
+    if compz in {"I", "V"}:
+        d_pteqr, e_pteqr, z_pteqr, work, info = pteqr(d, e, z=z[:-1], compz=compz)
+        assert_(info != 0)
```
@ilayn Ideally, this would raise a `ValueError`, too, right? How can we implement that?

mdhaber replied:
@ilayn even with:
```
<ftype> intent(in,out,copy), depend(n), dimension((compute_z=='N'?1:MAX(1,n)),(compute_z=='N'?0:n)) :: z
```
as you suggested, this still does not raise a `ValueError`; it behaves the same was as it did before.

ilayn commented:
My bad, I should have been clearer. This is one of the examples that I found randomly 

https://github.com/scipy/scipy/blob/master/scipy/linalg/flapack_gen.pyf.src#L209-L212

Here you see that the input is integer but resolves to a char based on the integer value.

ilayn commented:
> Do you agree that we also need some sort of check to ensure that z.shape[0] ==n when compute_z!='N'

Yes. Indeed dimension should have 
```
dimension((compute_z==0?1:MAX(1,n)),(compute_z==0?0:n))
``` 
such that when not required returns an empty array otherwise

mdhaber commented:
> My bad, I should have been clearer.

It was fine; I understood. I just disagreed : ) I decided to wait until the other issue was resolved in order to address it.

The convention seems wrong, and I think we should challenge it moving forward. One common reason to stick with a bad convention is that users expect the convention; they will assume the routine works like the rest and will be "astonished" when the new function doesn't work the same way. I don't think that applies here, as the existing convention requires that users either use trial and error or source code inspection to determine the mapping between the integers and characters. I doubt users will be more astonished to find that the syntax now follows the Fortran routine and that options are adequately documented by Netlib.

Are there better reasons to stick with the convention that do apply here?

ilayn commented:
One of the arguments is that you have to enter a capital letter and quotes instead exactly. Instead you can just type an integer. The other one is handling chars in the wrappers (well mostly fortran), it is terrible. In terms of UX, I have to agree that integers a bit better. But if we were to start from scratch I would have kept the letters. 

Consistent and a bit stupid API always beats an inconsistent API. Since we received no complaints on other places (and there are more of these quirks) I would argue that nobody cares.

I am -0.0 about this change. To be honest, f2py wrappers annoy me a lot so this bothers me the least. By the way, `sygst `and `hegst ` is even worse they have 1, 2, 3 as the `itype` arguments.

mdhaber commented:
OK. Changing.

mdhaber commented:
I'm baffled.  The wrapper is behaving as though `compute_z` is not being passed in as an `int` in the tests. It always fails `check((compute_z>=0) && (compute_z<=2))`, raising `ValueError`. If I remove this check, then `(compute_z?(compute_z==2?"I":"V"):"N")` always returns "V". Based on the error messages, it seems that the wrapper thinks it's being passed in as a string. e.g.,
```
E   _flapack.error: ((compute_z>=0) && (compute_z<=2)) failed for 1st keyword compute_z: spteqr:slen(compute_z)=1 compute_z="0"
```
What is going on?

Kai-Striega commented on scipy/linalg/flapack_pos_def_tri.pyf.src:
```diff
@@ -102,15 +102,14 @@ subroutine <prefix>pteqr(compute_z, n, d, e, z, ldz, work, info)
     callprotoargument char*, int*, <ctypereal>*, <ctypereal>*, <ctype>*, int*, <ctypereal>*, int*
-
-    character intent(in, optional), check(*compute_z=='N'||*compute_z=='V'||*compute_z=='I') :: compute_z='N'
+    character intent(in, optional), check((compute_z>=0) && (compute_z<=2)) :: compute_z = 0
```
> Based on the error messages, I would guess that the wrapper thinks it's a string. - @mdhaber

It does. You need to tell it to expect an integer **not** a character.

```suggestion
    integer intent(in, optional), check((compute_z>=0) && (compute_z<=2)) :: compute_z = 0
```

mdhaber replied:
Thanks. As you know, I'm not as familiar with these; I only did one very simple one.

Now I'm getting:
```
____________________________________________ test_pteqr[0-float32-float32] ____________________________________________
scipy\linalg\tests\test_lapack.py:2133: in test_pteqr
    d_pteqr, e_pteqr, z_pteqr, info = pteqr(d=d, e=e, z=z, compute_z=compute_z)
```
because:
```
<ftype> intent(in,out,copy), depend(n), dimension((compute_z==0?1:max(1,n)),(compute_z==0?0:n)) :: z
```
If `compute_z`==0, I think we don't care what the dimensions of `z` are; the documentation states `If  COMPZ = 'N', then Z is not referenced.` So I modified that line:
```
<ftype> intent(in,out,copy), depend(n), dimension((compute_z==0?shape(z, 0):max(1,n)),(compute_z==0?shape(z, 1):n)) :: z
```

Kai-Striega replied:
I'm not sure where you're getting errors from. All tests on the CI pass (the failing test is from differential evolution) as they do on my machine.

Kai-Striega commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -2074,12 +2074,145 @@ def test_pttrf_pttrs_NAG(d, e, d_expect, e_expect, b, x_expect):
+                           np.array([4.16, 5.25, 1.09, .62]),
+                           np.array([3.17, -.97, .55]),
+                           np.array([8.0023, 1.9926, 1.0014, 0.1237]),
+                           np.array([[0.6326,  0.6245, -0.4191,  0.1847],
```
Pep8 fixes suggested by the TravisCI
```suggestion
                         [(2,  # "I"
                           np.array([4.16, 5.25, 1.09, .62]),
                           np.array([3.17, -.97, .55]),
                           np.array([8.0023, 1.9926, 1.0014, 0.1237]),
                           np.array([[0.6326, 0.6245, -0.4191, 0.1847],
```

mdhaber replied:
Done.

ilayn commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -20,7 +20,7 @@
 
 from scipy.linalg import _flapack as flapack, lapack
-from scipy.linalg import inv, svd, cholesky, solve, ldl, norm, block_diag, qr
+from scipy.linalg import inv, svd, cholesky, solve, ldl, norm, block_diag, qr, eigh
```
PEP8 line too long

mdhaber replied:
OK

ilayn commented on scipy/linalg/flapack_pos_def_tri.pyf.src:
```diff
@@ -83,3 +83,34 @@ subroutine <prefix2c>pttrs(lower, n, nrhs, d, e, b, ldb, info)
+    ! high relative accuracy in the small eigenvalues of the original
+    ! matrix, if these eigenvalues range over many orders of magnitude.)
+    callstatement (*f2py_func)((compute_z?(compute_z==2?"I":"V"):"N"), &n, d, e, z, &ldz, work, &info)
+    callprotoargument char*, int*, <ctypereal>*, <ctypereal>*, <ctype>*, int*, <ctypereal>*, int*
```
char must be `int*`

Kai-Striega replied:
I'm sorry I think I misunderstood something here. Even though `compute_z` is (now) an integer it still resolves to be one of `'I'`,'`'V`' or '`N`', right? In which case it should be `char*`.

ilayn replied:
Oh yes true nevermind my comment. I need some coffee apparently

ilayn commented on scipy/linalg/flapack_pos_def_tri.pyf.src:
```diff
@@ -83,3 +83,34 @@ subroutine <prefix2c>pttrs(lower, n, nrhs, d, e, b, ldb, info)
+    callstatement (*f2py_func)((compute_z?(compute_z==2?"I":"V"):"N"), &n, d, e, z, &ldz, work, &info)
+    callprotoargument char*, int*, <ctypereal>*, <ctypereal>*, <ctype>*, int*, <ctypereal>*, int*
+    integer intent(in, optional), check((compute_z>=0) && (compute_z<=2)) :: compute_z = 0
+    integer intent(hide), check(n > 1), depend(d) :: n = len(d)
```
check is not needed here

ilayn replied:
Same deal with `e` dimension like we discussed in ptsvx wrappers.

mdhaber replied:
`dimension((n>0?n-1:0))`

mdhaber replied:
Removed check

mdhaber commented:
@ilayn I think we addressed all the comments. Azure couldn't find the build anymore, so I made a trivial push to re-trigger CI. If everything looks good, would you merge? Thanks for your help!

ilayn commented:
All green so clicking the button. Thanks all!
